demo/transient/fv_tpff90.F90

```diff
@@ -0,0 +1,334 @@
+
+    if (.not.mesh_file_flg) then
+       faces(1) = nx; faces(2) = ny; faces(3) = nz
+       lower(:) = 0.d0
+       upper(:) = 1.d0
+
+       call DMPlexCreateBoxMesh(PETSC_COMM_WORLD, dim, PETSC_FALSE, faces, lower, upper, &
+            PETSC_NULL_INTEGER, PETSC_TRUE, dm, ierr)
+       CHKERRA(ierr)
+    else
+       call DMPlexCreateFromFile(PETSC_COMM_WORLD, mesh_filename, PETSC_TRUE, dm, ierr)
+       CHKERRA(ierr)
+       call DMGetDimension(dm, dim, ierr)
+       CHKERRA(ierr)
+       if (dm_plex_extrude_layers > 0) then
+          call DMPlexExtrude(dm, PETSC_DETERMINE, -1.d0, PETSC_TRUE, PETSC_NULL_REAL, PETSC_TRUE, edm, ierr)
+          CHKERRA(ierr)
+          call DMDestroy(dm ,ierr)
+          dm = edm
+       end if
+    endif
+
+  end subroutine CreateDM
+
+end module snes_mpfaof90mod
+
+program main
+
+#include <petsc/finclude/petsc.h>
+#include <finclude/tdycore.h>
+#include <petsc/finclude/petscksp.h>
+#include <petsc/finclude/petscdm.h>
+#include <petsc/finclude/petscdmplex.h>
+#include <finclude/tdycore.h>
+
+  use tdycore
+  use petscvec
+  use petscdm
+  use petscdmplex
+  use petscdmplexdef
+  use petscksp
+  use petscsnes
+  use petscts
+  use snes_mpfaof90mod
+
+  implicit none
+
+  TDy                 :: tdy
+  DM                  :: dm
+  Vec                 :: U
+  SNES                :: snes
+  PetscFE             :: fe
+  PetscViewer         :: viewer
+  SNESConvergedReason :: reason
+
+  PetscInt            :: max_steps, step
+  PetscInt            :: step_mod
+
+  PetscReal           :: perm(9), resSat
+  PetscReal           :: dtime, ic_value
+
+  PetscBool           :: ic_file_flg, flg
+
+  PetscErrorCode      :: ierr
+  character (len=256) :: ic_filename
+  character(len=256)  :: string
+  PetscBool           :: use_seepage_bc
+
+  PetscInt, parameter :: successful_exit_code = 0
+  PetscInt, parameter :: unsuccessful_exit_code = 0
+
+  max_steps = 1;
+
+  call TDyInit(ierr); CHKERRA(ierr);
+  call TDyCreate(tdy, ierr); CHKERRA(ierr);
+
+  ! numerical method settings
+  call TDySetMode(tdy,RICHARDS,ierr)
+  call TDySetDiscretization(tdy, MPFA_O, ierr); CHKERRA(ierr);
+
+  ! command line settings
+  call PetscOptionsGetString(PETSC_NULL_OPTIONS,PETSC_NULL_CHARACTER,"-ic_filename", ic_filename, ic_file_flg,ierr); CHKERRA(ierr);
+  call PetscOptionsGetInt(PETSC_NULL_OPTIONS,PETSC_NULL_CHARACTER,'-max_steps',max_steps,flg,ierr); CHKERRA(ierr);
+  call PetscOptionsGetBool(PETSC_NULL_OPTIONS,PETSC_NULL_CHARACTER,"-use_seepage_bc",use_seepage_bc,flg,ierr); CHKERRA(ierr);
+
+  ! Set a constructor for a DM.
+  call TDySetDMConstructor(tdy, CreateDM,ierr); CHKERRA(ierr)
+
+  ! Apply overrides.
+  call TDySetFromOptions(tdy,ierr); CHKERRA(ierr)
+
+  call TDyGetDM(tdy, dm, ierr); CHKERRA(ierr)
+  call DMGetDimension(dm, dim, ierr);  CHKERRA(ierr)
+  call PetscFECreateDefault(PETSC_COMM_SELF, dim, 1, PETSC_FALSE, "p_", -1, fe, ierr); CHKERRA(ierr)
+  call PetscObjectSetName(fe, "p", ierr); CHKERRA(ierr)
```

Comment:
Is this `p` field the MPI rank of the process? Does our model assume this field exists? If so, maybe we could fold this code into `TDySetup`.